### PR TITLE
[Core] Mix per-prompt seeds into seeded SamplingParams in the offline API

### LINKED ISSUE #50440: [Bug]: Broadcasting one seeded SamplingParams across a batch gives every prompt the same noise vector
### Summary

When a batch of prompts is submitted to the offline API with one seeded `SamplingParams` (`LLM.generate(prompts, sampling_params)` / `LLM.chat`), vLLM broadcasts the same object — and therefore the same seed — to every request in the batch (`OfflineInferenceMixin._params_to_seq` returns `[params] * num_requests`). All trajectories then draw an identical noise vector: their sampled tokens are correlated, not independent draws from `softmax(logits)`. Over multiple trajectories, this biases the marginal token distribution rather than averaging out.

Scope notes:

- `n>1` within a request is already handled: `ParentRequest._get_child_sampling_params` derives `seed + index` per child.
- Passing an explicit `Sequence[SamplingParams]` whose seeds already differ avoids the correlation, but the natural broadcast path silently produces it. An explicit list that repeats the same seed hits the bug just the same.
- The two call forms are also inconsistent: `generate([a], SamplingParams(seed=42))` and `generate([a], [SamplingParams(seed=42)])` are the same request expressed two ways, and any per-request seed derivation has to apply to both or they will diverge.
- A fix should derive each request's seed from `(seed, prompt)` rather than from batch position, so a prompt's result does not depend on batch composition or order.

### Affected code

`vllm/entrypoints/offline_utils.py` — `_params_to_seq` broadcasts one `SamplingParams` object (one seed) to all requests:

```python
return [params] * num_requests
```

`vllm/v1/sample/ops/topk_topp_sampler.py` — where the shared seed becomes a shared noise vector:

- `random_sample()` — per-request generator loop:

```python
q[i].exponential_(generator=generator)   # one Exp(1) per (request, token_id)
```

- `sample_with_exponential_noise()` — the argmax:

```python
probs.div_(q).argmax(dim=-1).view(-1)
```

### Mechanism

Random sampling uses the Gumbel-max trick in the probability domain: draw one `Exp(1)` variable `q` per token, then take `argmax(probs / q)` — algebraically `argmax(log p + g)` with `g = -log(q) ~ Gumbel(0,1)`.

Each request initializes its own `torch.Generator` from the integer `seed` **once at request entry**, reused across decode steps (prefill consumes no RNG). Consequently, at step 1:

- Every request with `seed=N` has the same generator state.
- Each request draws a vocab-sized exponential vector `q` that is identical across all same-seed requests.
- The argmax depends on `log p - log q`. The signal `log p` is per-prompt; the noise `q` is shared across all same-seed requests.

### Expected vs actual

- **Expected:** broadcasting a seeded `SamplingParams` gives each prompt an independent noise stream, derived deterministically from `(seed, prompt)` so results are reproducible and independent of batch composition.
- **Actual:** every prompt in the batch receives the identical seed and therefore the identical noise vector.

### PR DESCRIPTION
**Pull request by tolleybot**
## Purpose

Fixes #50440.

When requests in an offline-API batch (`LLM.generate` / `LLM.chat`) share a `SamplingParams` seed, every request draws an identical Exp(1) noise vector in `random_sample`, so sampled tokens are correlated across the whole batch: any token the shared noise vector suppresses (or favors) is suppressed/favored for every prompt simultaneously, biasing the marginal token distribution over the population rather than averaging out.

This PR mixes each seeded request's seed with its **prompt content**, after params are expanded to one per prompt:

```
request_seed = (seed + sha256_digest(prompt_content)[:8]) % 2**63
```

- **Uniform across call forms**: mixing applies after expansion, so `generate([a], SamplingParams(seed=42))` and `generate([a], [SamplingParams(seed=42)])` give identical results.
- **Content-based, not index-based**: a prompt's derived seed — and therefore its output — depends only on the `(seed, prompt)` pair, independent of batch composition and order. Re-running any prompt alone reproduces its in-batch result. (An index-based scheme like `seed + i` would reintroduce a dependence on batch position, the same class of problem as #17524.)
- **Same (seed, prompt) → same result, always**: repeated identical prompts in a batch share their noise by construction. Drawing multiple distinct samples of one prompt is what `n>1` is for — `ParentRequest._get_child_sampling_params` derives distinct per-child seeds (`seed + child_index`) within each request.
- The digest tags text and token-ID prompts with distinct type bytes (no cross-type collisions) and hashes multi-modal items by decoded content where possible — PIL image pixels (encoding-independent) and array buffers — falling back to type names for anything else. Prompts with the same text but different images therefore derive different seeds. Residual limitation: multi-modal items that expose neither pixels nor an array buffer reduce to their type name and can still share a digest.

Behavior notes:
- Same call → same results: full determinism is preserved.
- Unseeded and pooling params keep the existing shared-object fast path; the OpenAI-compatible server path is untouched (it builds per-request params).
- **Outputs of seeded offline generation change relative to previous releases** (including single-prompt calls, since the seed is now mixed with the prompt digest). The previous behavior — one shared noise stream for the whole batch — is exactly the bug being fixed; any golden outputs recorded under a seed will need re-recording.
- Deliberately giving *different* prompts identical noise (common-random-numbers comparisons) is no longer possible; that capability was incidental, undocumented, and is the same mechanism that produces the population-level bias.

Duplicate-work check (per AGENTS.md): `gh issue view 50440 --comments` shows no assigned work and no linked PR; `gh pr list --state open` searches for "50440 in:body", "seed trajectory sampling", and "per-request seed generator" return no PR addressing this. The only related merged work is the `n>1` child-seed derivation (#2514 lineage), which this PR complements at the batch level.

## Test Plan

New unit tests added to the existing `tests/entrypoints/unit_tests/test_offline_utils.py` (CPU-only, no GPU required; the 4 pre-existing validation tests in that file still pass):

```bash
pytest tests/entrypoints/unit_tests/test_offline_utils.py -v
```

Covers: mixed seeds are unique per prompt, deterministic, and don't mutate the caller's object; single-params and explicit-list forms produce identical seeds; batch-composition/order independence (reorder + singleton subset reproduce the same per-prompt seeds); repeated identical prompts share their derived seed; token-ids digests; image-content digests (identical pixels share a seed, different pixels don't); unseeded `SamplingParams` and `PoolingParams` pass through unchanged.

Lint: `pre-commit run --files vllm/entrypoints/offline_utils.py vllm/entrypoints/llm.py tests/entrypoints/unit_tests/test_offline_utils.py` (includes mypy) — all hooks pass.

## Test Result

```
tests/entrypoints/unit_tests/test_offline_utils.py::test_rejects_mismatched_params PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_rejects_mismatched_lora_requests PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_rejects_mismatched_priority PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_matching_lengths_pass_through PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_mixed_seeds_unique_and_deterministic PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_single_params_and_list_give_identical_seeds PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_mixed_seed_independent_of_batch_composition PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_duplicate_prompts_share_seed PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_mixed_seed_uses_token_ids_fingerprint PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_mixed_seed_uses_image_content PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_unseeded_sampling_params_pass_through_unchanged PASSED
tests/entrypoints/unit_tests/test_offline_utils.py::test_pooling_params_pass_through_unchanged PASSED

12 passed
```

End-to-end comparison (CPU backend, `facebook/opt-125m`, source build; re-verified after rebase onto current `main`): 64 distinct prompts (`"Item {i}:"`), one shared `SamplingParams(seed=1234, temperature=2.0, max_tokens=1)`:

| | `main` | this PR |
|---|---|---|
| distinct first sampled tokens across the 64 prompts | **1** (all 64 prompts sampled token 11944) | **64** (independent draws) |
| identical results on rerun of the same call | yes | yes |

On `main` the shared noise vector dominates the near-identical next-token distributions and every prompt in the batch samples the same token — the population-level collapse described in #50440. With this PR the draws are independent while remaining fully deterministic.

---

AI assistance disclosure: this change was developed with AI assistance (Claude Code). The submitting human has reviewed every changed line and run the tests above. Model quality evals beyond the end-to-end comparison above were not run: the change does not alter model forward passes or the sampler kernels — it only decorrelates the RNG seed assignment across batched seeded requests, and greedy/unseeded paths are untouched.

### CODE REVIEW

**Comment by stackedsax** [comment]
Hey there, curious if anyone has looked at this one yet.  It would be great to know whether others think this would be useful.
